src/rollback_to_stable/rts_btree.c

```diff
@@ -92,6 +92,18 @@ __rts_btree_abort_update(WT_SESSION_IMPL *session, WT_ITEM *key, WT_UPDATE *firs
      * store, it will have a different stop time point.
      */
     if (stable_upd != NULL) {
+        /*
+         * During recovery, there shouldn't be any updates in the update chain except when the
+         * updates are from a prepared transaction. Reset the transaction ID of the stable update
+         * that was restored as part of the unstable prepared tombstone.
+         */
+        if (F_ISSET(S2C(session), WT_CONN_RECOVERING)) {
+            WT_ASSERT(session, stable_upd->next == NULL);
+            WT_ASSERT(session,
+              stable_upd->txnid == WT_TXN_NONE || F_ISSET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS));
```

Comment:
Can we fix the flag as well?

Reply:
@quchenhao What do you mean by fix the flag?

Reply:
when reading the update from the disk we set it as WT_UPDATE_PREPARE_RESTORED_FROM_DS but the update is not a prepared update.

Reply:
I think this should be a different ticket. Instead of fixing it here, we should find where it is wrongly set.

Reply:
Yes, I mean we should fix where it is set.

Reply:
@quchenhao @etienneptl  It's not clear to me where we went wrong setting this flag. This flag is only set from the `__page_inmem_prepare_update` function so I'm not sure how a non-prepared update got it.

Reply:
Do we copy the flags from one update to the other?

Reply:
```
    WT_RET(__wt_upd_alloc(session, value, WT_UPDATE_STANDARD, &upd, &size));
    total_size += size;
    upd->durable_ts = unpack->tw.durable_start_ts;
    upd->start_ts = unpack->tw.start_ts;
    upd->txnid = unpack->tw.start_txn;
    F_SET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS);
```

The issue is here we always mark the onpage value as WT_UPDATE_PREPARE_RESTORED_FROM_DS in `__page_inmem_prepare_update `. We should only do that if there is no stop or both the start and the stop are prepared updates.

Reply:
We should do something like this:

```
diff --git a/src/btree/bt_page.c b/src/btree/bt_page.c
index 5bd5b9f25..7a4d7b10e 100644
--- a/src/btree/bt_page.c
+++ b/src/btree/bt_page.c
@@ -144,7 +144,6 @@ __page_inmem_prepare_update(WT_SESSION_IMPL *session, WT_ITEM *value, WT_CELL_UN
     upd->durable_ts = unpack->tw.durable_start_ts;
     upd->start_ts = unpack->tw.start_ts;
     upd->txnid = unpack->tw.start_txn;
-    F_SET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS);

     /*
      * Instantiate both update and tombstone if the prepared update is a tombstone. This is required
@@ -170,13 +169,16 @@ __page_inmem_prepare_update(WT_SESSION_IMPL *session, WT_ITEM *value, WT_CELL_UN
           unpack->tw.start_txn == unpack->tw.stop_txn) {
             upd->durable_ts = WT_TS_NONE;
             upd->prepare_state = WT_PREPARE_INPROGRESS;
-        }
+            F_SET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS);
+        } else
+            F_SET(upd, WT_UPDATE_RESTORED_FROM_DS);

         tombstone->next = upd;
         *updp = tombstone;
     } else {
         upd->durable_ts = WT_TS_NONE;
         upd->prepare_state = WT_PREPARE_INPROGRESS;
+        F_SET(upd, WT_UPDATE_PREPARE_RESTORED_FROM_DS);
         *updp = upd;
     }
```

Reply:
Fix was done has part of [this commit](https://github.com/wiredtiger/wiredtiger/pull/10864/commits/2ad4f4b203fb71661460698805109664373fb514).